### PR TITLE
[CPDLP-1691] Fetch partnerships via API v3

### PR DESCRIPTION
**Pull request by leandroalemao**
### Context

Providers can view partnerships via API v3

- Ticket: [CPDLP-1691](https://dfedigital.atlassian.net/browse/CPDLP-1691)

### Changes proposed in this pull request
Add a new endpoint GET: `/api/v3/partnerships/ecf`

### Guidance to review
[API-v3-Reference](https://ecf-dev.london.cloudapps.digital/api-reference/reference-v3.html#api-v3-partnerships-ecf-get)
[API v3 changes](https://docs.google.com/document/d/1AfOoSCCZKIdvFL6sijbi7vmsfZNu7RLHejdAWEUTmqo/edit#heading=h.v96zs25e5q5l)

### CODE REVIEW

File: app/serializers/api/v3/ecf_partnership_serializer.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+require "jsonapi/serializer/instrumentation"
+
+module Api
+  module V3
+    class ECFPartnershipSerializer
+      include JSONAPI::Serializer
+      include JSONAPI::Serializer::Instrumentation
+
+      set_id :id
+      set_type :partnership
```
Comment (StupidCodeFactory) [inline app/serializers/api/v3/ecf_partnership_serializer.rb:12]:
```suggestion
      set_type "partnership-confirmation"
```

File: app/controllers/concerns/api_orderable.rb [hunk 2]
```diff
@@ -3,7 +3,15 @@
...
-private
+  module ClassMethods
+    attr_reader :model_klass
+
+  private
+
+    def orderable(model_klass: nil)
+      @model_klass = model_klass
+    end
+  end
```
Comment (StupidCodeFactory) [inline app/controllers/concerns/api_orderable.rb:14]: Let's avoid unnecessary meta-programming. I'd rather the class to injected into the `#sort_params` method.

```ruby
def sort_params(params, model: controller_name.classify.constantize)
  # ...
end
```
and in the controller
```ruby
def ecf_partnerships
  @ecf_partnerships ||= ecf_partnerships_query.partnerships.order(sort_params(params, model: Api::V3::ECFPartnershipSerializer))
end
```

File: app/models/partnership.rb [hunk 3]
```diff
@@ -51,6 +51,11 @@ def active?
     challenged_at.nil? && challenge_reason.nil? && pending == false
   end
 
+  def status
+    return "active" if active?
+    return "challenged" if challenged?
+  end
```
Comment (StupidCodeFactory) [inline app/models/partnership.rb:57]: ✂️ this is a presentational concern and should be moved to the serializer.

File: app/services/api/v3/ecf_partnerships_query.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      attr_reader :lead_provider, :params
+
+      def initialize(lead_provider:, params:)
+        @lead_provider = lead_provider
+        @params = params
+      end
+
+      def partnerships
+        scope = lead_provider.partnerships.includes(:school, :cohort, :delivery_partner)
+        scope = scope.where("partnerships.cohort_id IN (?)", with_cohorts.map(&:id)) if filter[:cohort].present?
```
Comment (StupidCodeFactory) [inline app/services/api/v3/ecf_partnerships_query.rb:15]: I'd prefer avoiding hardcoding the table's name in the query and let ActiveRecord so its thing as much as possible.
```ruby
scope.where(partnerships: { cohort:  with_cohorts} ) if filter[:cohort].present?
scope.where(partnerships: { delivery_partner: [delivery_partner_id_filter]} if delivery_partner_id_filter.present?
```

File: app/services/api/v3/ecf_partnerships_query.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      end
+
+    private
+
+      def filter
+        params[:filter] ||= {}
+      end
+
+      def delivery_partner_id_filter
+        filter[:delivery_partner_id]&.split(",")
```
Comment (StupidCodeFactory) [inline app/services/api/v3/ecf_partnerships_query.rb:29]:
```suggestion
        filter[:delivery_partner_id].split(",")
```

File: app/controllers/api/v3/ecf_partnerships_controller.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+module Api
+  module V3
+    class ECFPartnershipsController < Api::ApiController
```
Comment (StupidCodeFactory) [inline app/controllers/api/v3/ecf_partnerships_controller.rb:5]: The pattern we're following for namespacing is:
```suggestion
module Api
  module V3
    module ECF
      class PartnershipsController < Api::ApiController
    end
  end
end
```

File: app/controllers/api/v3/ecf_partnerships_controller.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+      def ecf_partnerships_query
+        Api::V3::ECFPartnershipsQuery.new(
+          lead_provider:,
+          params: ecf_partnership_params,
+        )
+      end
+
+      def ecf_partnership_params
+        params.permit(:sort, filter: %i[cohort updated_since delivery_partner_id])
```
Comment (StupidCodeFactory) [inline app/controllers/api/v3/ecf_partnerships_controller.rb:39]:
```suggestion
        params
          .with_defaults({sort: "", filter: {delivery_partner_id: "", updated_since: "", cohort: ""}})
          .permit(:sort, filter: %i[cohort updated_since delivery_partner_id])
```

File: app/serializers/api/v3/ecf_partnership_serializer.rb (hunk 1 above)
Comment (StupidCodeFactory) [inline app/serializers/api/v3/ecf_partnership_serializer.rb:12]:
```suggestion
      set_type "partnership-confirmation"
```

**Comment by leandroalemao** [comment]
Hi @StupidCodeFactory tks for your review. All comments have been addressed now. Tkss 👍 😄

**Pull request merged**